### PR TITLE
[dynamo][guard overhead] Data ptr guard optimizer state tensors

### PR DESCRIPTION
**Pull request by mlazos**
Stricter (but faster) guarding on optimizer state tensors

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by anijain2305 — APPROVED** [review]

File: torch/_dynamo/variables/optimizer.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ def __init__(
                 group["capturable"] = True
 
             for p in group["params"]:
-                mark_static_address(p, guard=False)
```
Comment (mlazos) [inline torch/_dynamo/variables/optimizer.py:73]: @anijain2305 I needed this

File: torch/_dynamo/variables/optimizer.py [hunk 2]
```diff
@@ -166,8 +166,13 @@ def map_sources_and_install_guards(self, tx):
...
+        # Note: we don't mark complex params as static addresses
+        # The main reason for this is that both a real view
+        # and the original complex tensor will share their data_ptr
+        # so we should guard on the dtype in this case.
         def mark_static(x):
-            mark_static_address(x, guard=False)
+            if not x.is_complex():
+                mark_static_address(x)
 
         tree_map_only(torch.Tensor, mark_static, self.value.state)
```
Comment (anijain2305) [inline torch/_dynamo/variables/optimizer.py:177]: But we do this here already, @mlazos?

Comment (anijain2305) [reply torch/_dynamo/variables/optimizer.py:177]: Maybe move this to init as well?

Comment (mlazos) [reply torch/_dynamo/variables/optimizer.py:176]: Discussed offline, treemap doesn't iterate over keys :(

File: torch/_dynamo/output_graph.py [hunk 3]
```diff
@@ -723,11 +723,10 @@ def register_attr_or_module(
                 # are registered as get_attr nodes in the root graph.
                 tracer = self.root_tracer
 
-            if not is_constant_source(source):
-                install_guard(source.make_guard(GuardBuilder.TENSOR_MATCH))
-
             if get_static_address_type(target) == "guarded":
-                install_guard(source.make_guard(GuardBuilder.DATA_PTR_MATCH))
+                install_guard(source.make_guard(GuardBuilder.ID_MATCH))
```
Comment (anijain2305) [inline torch/_dynamo/output_graph.py:727]: Suggest to make it DATA_PTR_MATCH. And just add a line to DATA_PTR in guards.py


```
# C++ guard for DATA_PTR_MATCH already checks the type internally
if not config.enable_cpp_guard_manager:
     self.TYPE_MATCH(guard)
```

**Pull request closed without merging**